Add secure-pod-defaults flag to default Pods to 'restricted' profile by default

evankanderson opened the pull request:
Fixes #13395

Adds a feature flag (deactivated for the current release) which defaults launched pods to a mostly-secure version of the `restricted` profile that shouldn't be too onerous. (i.e. it does not set `runAsNonRoot`, because Docker defaults to building containers that run as root)

## Proposed Changes

* Adds a feature flag `secure-pod-defaults`, which is currently `Disabled`
* When `secure-pod-defaults` is set, it will fill in _unset_ container `SecurityContext`s with secure defaults
* When the flag is set, if specific insecure SecurityContext features are requested, they will be retained.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Adds the `secure-pod-defaults` feature, which is defaulted to Disabled in
this release.

When enabled, containers described by users will have best-practice
SecurityContext features enabled unless insecure settings are specifically
requested.
```

evankanderson commented:
/assign @dprotaso @mattmoor @psschwei 

I realize this is very last-minute, but I'd love to get this into 1.8 on Tuesday, so that we can move the flag to Beta (`Enabled` by default) for 1.9.

mattmoor commented on pkg/apis/serving/v1/revision_defaults.go:
```diff
@@ -158,6 +162,37 @@ func (*RevisionSpec) applyProbes(container *corev1.Container) {
+	if container.SecurityContext.Capabilities.Drop == nil {
+		container.SecurityContext.Capabilities.Drop = []corev1.Capability{"ALL"}
+	}
+	// Default in NET_BIND_SERVICE to allow binding to low-numbered ports.
+	if container.SecurityContext.Capabilities.Add == nil {
+		container.SecurityContext.Capabilities.Add = []corev1.Capability{"NET_BIND_SERVICE"}
+	}
```
I wonder if this will trip any naive policies that look for ~any added capabilities.  We should only need this when the user is specifically requesting a low port, which we could know from the enclosing context, but it'd take some gross plumbing to figure that out.

evankanderson replied:
Updated to do this check.

mattmoor commented on pkg/apis/serving/v1/revision_defaults.go:
```diff
@@ -158,6 +162,37 @@ func (*RevisionSpec) applyProbes(container *corev1.Container) {
+	if container.SecurityContext.AllowPrivilegeEscalation == nil {
+		container.SecurityContext.AllowPrivilegeEscalation = ptr.Bool(false)
+	}
+	if container.SecurityContext.SeccompProfile == nil {
+		container.SecurityContext.SeccompProfile = &corev1.SeccompProfile{}
+	}
+	if container.SecurityContext.SeccompProfile.Type == "" {
+		container.SecurityContext.SeccompProfile.Type = corev1.SeccompProfileTypeRuntimeDefault
+	}
```
We don't allow `SeccompProfile` thru the fieldmask today (it looks like it may be newer than our field triage because it isn't copied or explicitly cleared): https://github.com/knative/serving/blob/a18077cd9bbfe9e6208b2202fc39dfa3c79fc837/pkg/apis/serving/fieldmask.go#L623-L639

So my main concern here would be that we are changing a default value that we don't allow folks to override.

evankanderson replied:
Good point, I can make a PR to make this settable.

evankanderson replied:
#13401

mattmoor reviewed (commented):
Generally I think this is a positive change.  Most of my feedback is focused on a future where this defaults on, ways it might break folks (directly or indirectly), and whether folks have the tools to workaround any breakage.

FWIW, we run all of our services with the following today:
```
        securityContext:
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          capabilities:
            drop:
            - all
```
... and I'm looking forward to adding `allowPrivilegeEscalation: false` to this.  As I said inline, I don't think we allow specifying seccomp profile today, which we should maybe allow? 🤔 

If we wanted to encourage a better default security posture for folks, we could consider using the new warning capabilities in `apis.FieldError` to encourage folks to explicitly set security features (well in advance of a breaking default change, in line with our versioning/deprecation policies).  We're using warnings in sigstore/policy-controller downstream and they are pretty neat!

psschwei commented:
Will give a full review on Monday, but wonder if there's any overlap with #13376

evankanderson commented:
I wrote #13399 for validation to warn about the insecure defaults.

skonto commented:
> if there's any overlap with https://github.com/knative/serving/pull/13376
Regarding overlapping with https://github.com/knative/serving/pull/13376 we can just do all required changes in this PR here.

Btw I left a comment [here](https://github.com/knative/serving/pull/13376#issuecomment-1280555176). I am in favor of using defaults that are more secure, avoid psp warnings, but also allow users to unset stuff on demand assuming they know what their doing.
It depends on our strategy, we can be as opinionated we want as long as users can change that and they know about it.
The downside is that users might expect whatever K8s has as a default (breaks UX), but we have broken UX already by disabling certain stuff with several flags.

skonto commented on pkg/apis/serving/v1/revision_defaults.go:
```diff
@@ -158,6 +162,56 @@ func (*RevisionSpec) applyProbes(container *corev1.Container) {
+		updatedSC.Capabilities = &corev1.Capabilities{}
+	}
+	if updatedSC.Capabilities.Drop == nil {
+		updatedSC.Capabilities.Drop = []corev1.Capability{"ALL"}
```
If `updatedSC.AllowPrivilegeEscalation` is nil and we set it to false and later on we have `updatedSC.Capabilities.Drop != nil` with `CAP_SYS_ADMIN` present  would that be a problem (https://github.com/kubernetes/website/issues/30104)?

evankanderson replied:
Per that note, I don't think we allow `privileged` to be set, so I don't think we get into the forcing issue.

And yes, the idea is that:

1. The flag defaults to false or can be flipped to false, so users can avoid things altogether for a while.

2. The flag only causes the _default_ values to be filled in, so there are ways to explicitly request the old behavior if needed.

skonto replied:
Don't we allow updatedSC.AllowPrivilegeEscalation set via https://github.com/knative/serving/pull/13395?

psschwei replied:
> Per that note, I don't think we allow `privileged` to be set, so I don't think we get into the forcing issue.

That's right, `SecurityContext.privileged` isn't allowed, though because a _pod's_ security context is behind a feature gate, it fails silently vs. getting caught in the validation as in a _container's_ security context.

> with CAP_SYS_ADMIN present, would that be a problem or K8s 

Since adding capabilities requires enabling a feature gate, we could put extra warnings in the release notes / documentation in for anyone who is flipping that gate on to further mitigate any risk.

evankanderson commented:
This is one of those "I reserve the right to wake up smarter" moments. I realized thanks to Matt's PR that all the SecurityContext values were pointer-ish, so it was possible to distinguish between "I don't care about security" and "I need to do something dangerous". That tipped me into the "we should do something safer than the Kubernetes default, even though the experience is different".

psschwei commented on pkg/apis/serving/v1/revision_defaults.go:
```diff
@@ -158,6 +162,56 @@ func (*RevisionSpec) applyProbes(container *corev1.Container) {
+	if updatedSC.AllowPrivilegeEscalation == nil {
+		updatedSC.AllowPrivilegeEscalation = ptr.Bool(false)
+	}
+	if psc.SeccompProfile == nil || psc.SeccompProfile.Type == "" {
+		if updatedSC.SeccompProfile == nil {
+			updatedSC.SeccompProfile = &corev1.SeccompProfile{}
+		}
+		if updatedSC.SeccompProfile.Type == "" {
+			updatedSC.SeccompProfile.Type = corev1.SeccompProfileTypeRuntimeDefault
+		}
+	}
```
#13401 needs to merge for this not to feel the wrath of the validating webhook...

psschwei replied:
in that vein, it probably wouldn't be a bad idea to have an e2e test for this before we default to enabling this (see also the issue above about needed to enable another feature to get this working)

psschwei commented on config/core/configmaps/features.yaml:
```diff
@@ -40,6 +40,13 @@ data:
+    #
+    # This value will default to "enabled" in a future release,
+    # probably Knative 1.8
+    secure-pod-defaults: "disabled"
```
I also ran into an issue where I was getting a validation error on seccompProfile, even after setting this to enabled and pulling in the other patch. Enabling the pod security context feature fixed it, but I wonder if we need to consolidate some of the behaviors... at the very least, maybe using secure defaults should turn on pod security contexts also (though something about that feels a little hacky...)

evankanderson replied:
I think you're correct about having this feature enable the other one. That actually seems reasonable to me, and at some point if we make this field default-true or even always-enabled, we can retire the other flag, too, possibly.

evankanderson commented:
/reopen

oops, that's not what I meant to do.

dprotaso commented:
Note - K8s provides a knob for setting the default seccompProfile for all workloads (to RuntimeDefault)

https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads

I wonder if we should skip setting this property since it's breaking OpenShift users and people using gVisor (https://github.com/knative/serving/issues/13471)

dprotaso commented:
Looks like long-term k8s will default to the `RuntimeDefault` profile as well

https://github.com/kubernetes/enhancements/tree/9a124fd29d1f9ddf2ff455c49a630e3181992c25/keps/sig-node/2413-seccomp-by-default#beta-to-ga-graduation

dprotaso commented:
Following up from the discussion on #13471  what would be useful here is for the feature flag to allow operators to customize the seccompProfile.type.

Since gVisor users would want to opt into the better security posture but not set the seccomp profile.

skonto commented:
We should have something that allows to switch on full compliance and also configure any parts of it.

mattmoor commented on config/core/configmaps/features.yaml:
```diff
@@ -40,6 +40,13 @@ data:
     # this example block and unindented to be in the data block
     # to actually change the configuration.
 
+    # Default SecurityContext settings to secure-by-default values
+    # if unset.
```
I'm curious what the compatibility guarantees of this are.  Personally, I'd love for this to evolve and _**change**_ as the best practices change, but that means minor version bumps could be breaking!  It would be good to link to some guidance we expect this flag to follow, so it seems less ad hoc.  Even a knative docs page would be better than nothing.

I suspect that stuff like disallowing privilege escalation and running as non-root as non-contentious parts of this, but I'm curious about potentially more contentious things like read-only root filesystems?

skonto replied:
Some projects consider seccompProfile possible risk https://github.com/istio/istio/issues/35894.

evankanderson replied:
I think we'll want to evolve and change this as well, assuming that Kubernetes continues to add "turn this knob to be secure" settings. One interesting challenge will be supporting new security settings across a range of Kubernetes API versions -- we may need to be more sophisticated about recognizing what fields need to be set for security.

I imagine if we end up needing to add more fields, we'll want to add new feature flags for those fields, e.g. `pod-security-defaults-k8s-1-27`, etc.

I'm also hoping / wishing that Kubernetes would pivot to a "secure by default" posture, but I understand that they have a lot of historical debt to pay down there (see `seccempProfile`, for example).

dprotaso commented:
FYI: I opened a gVisor issue to allow `RuntimeDefault` as an option - https://github.com/google/gvisor/issues/8248

skonto commented:
@evankanderson @psschwei
I  am testing this PR downstream so faced a couple of issues given that the intention was to enable this by default in 1.9:

a) Queue proxy requires similar code
b) Adding new defaults suffer from https://github.com/knative/serving/issues/11512, so things brake during upgrades if users turn this on.
c) Setting seccompProfile to RuntimeDefault may not work depending on the platform, on OCP for example I noticed that it creates problems with `TestShouldRunAsUserContainerDefault` as container will not run with the image uid.

evankanderson commented:
I'm going to work on this and the warning one in the next week or so, thanks for the testing. (I know OpenShift in particular can be tricky -- any tips for a test environment there? Is there a recommended local / single VM environment that's sufficient for Knative?)

skonto commented:
@evankanderson you could try [crc](https://crc.dev/crc/#installing_gsg) (OCP 4.11.13, K8s 1.24), in addition I could test the updated PR on a real cluster. Btw it would be useful to label a ksvc with a special key, pair so that we can avoid setting seccompProfile if that label is set to "true". This was we can cover corner cases where seccompProfile creates issues depending on the platform definition of the RuntimeDefault. 
More info on how PSS (scc V2) is implemented on OCP can be found [here](https://github.com/redhat-openshift-ecosystem/community-operators-prod/discussions/1417) and the key takeaways are:

> V2 does not permit allowPrivilegeEscalation=true
** Empty or false is compatible with v1 SCC and therefore works on OCP versions < 4.11
V2 requires you to leave the dropped capabilities empty, set it to ALL, or add only NET_BIND_SERVICE
** By being accepted as v2 the SCC will always drop ALL. V1 only dropped KILL, MKNOD, SETUID, SETGID capabilities.
V2 still allows explicitly adding the NET_BIND_SERVICE capability
V2 requires you to either leave SeccompProfile empty or set it to runtime/default
** Empty is compatible with v1 and works on OCP versions < 4.11

so it is indeed tricky and it would be good to allow SeccompProfile to be empty per service if user needs to.

evankanderson commented:
This will depend on #13565 for defaulting in the BYO-name case; I'll merge that in later today, along with enabling the SecurityContext feature when this feature is enabled.

evankanderson commented:
In addition to the dependency on #13565, I've added a the ability to explicitly request the Kubernetes defaults for `SeccompProfile` (at the PodSpec level) and `Capabilities` and `AllowPrivilegeEscalation` (both at the container level, and already allowed).

psschwei commented:
don't think we wanted to close this...

evankanderson commented:
> In addition to the dependency on #13565, I've added a the ability to explicitly request the Kubernetes defaults for `SeccompProfile` (at the PodSpec level) and `Capabilities` and `AllowPrivilegeEscalation` (both at the container level, and already allowed).

I'd originally written this so that you could request the default (`Unconfined`) behavior at the Pod level, but not the safe default (`RuntimeDefault`). I've changed it so you can request either value when the flag is enabled, so you can explicitly opt in to the safe behavior if desired.

evankanderson commented:
(Tests are now passing, though that could be because I've got the behavior behind a flag.)

I'll try to get an integration test or two in for this, but I'd really like to get the flag in **DEFAULT OFF** for Knative 1.9, so we can talk about flipping the flag to on in 1.10 in April.

evankanderson commented:
It turns out `capabilities.drop` is marked `omitempty`, so I can't distinguish between `nil` and `[]`; I've adjusted the code to default in capabilities only if entire object is empty.

evankanderson commented:
@skonto @psschwei @dprotaso 

I think this is ready for review again

psschwei commented on pkg/apis/serving/fieldmask_test.go:
```diff
@@ -728,6 +728,51 @@ func TestPodSecurityContextMask_FeatureEnabled(t *testing.T) {
+		&config.Config{
+			Features: &config.Features{
+				SecurePodDefaults:      config.Enabled,
+				PodSpecSecurityContext: config.Disabled,
```
this currently defaults to `disabled`, so should we remove it from the test? (that would keep us from ever getting caught up if it defaults to `enabled` in the future and there's a behavior change)

evankanderson replied:
I wanted to validate that users who have `PodSpecSecurityContext` disabled can still specifically set the fields needed to disabled the defauts set in `SecurePodDefaults`.

psschwei reviewed (commented):
nit on one of the tests, but otherwise lgtm

psschwei reviewed (commented):
/lgtm

/hold
to give @skonto @dprotaso @mattmoor a chance to comment (otherwise will unhold in the morning before cutting the release)

skonto commented:
Some questions:
a) Regarding runAsNonRoot is it that we might hit some image that is built with root and we will stop it from running if we set runAsNonRoot=true?
b) Users will have to enable `kubernetes.containerspec-addcapabilities` if they add some port < 1024 correct? I guess we don't test that part. Downstream some tests fail (grpc ping) not sure yet, will have to investigate.

evankanderson commented:
Yeah, I didn't touch `runAsNonRoot` for now because Docker defaults to building images with UID=0, and some popular images such as nginx may be written to run as UID 0.  I'd love to fix this, but I want another cycle or a separate knob for it.

evankanderson commented:
For (b), it's intended that we will set `NET_BIND_SERVICE` to enable ports < 1024.  It's possible that isn't sufficient for some cases -- that's one reason I'm leaving the default at `false` for right now (also, I'm not a monster that jumps straight to beta, but mostly I'm worried about ironing out these cases with real tests).

If you have an example image that doesn't work with port <1024 and otherwise does work, it will shorten my repro efforts.

evankanderson commented:
(I'm going to cancel hold around 7-8pm, so this can get one round of nightlies before going into the release. I should have done it earlier, but... nothing like a deadline to sharpen the will.)

evankanderson commented:
/hold cancel

evankanderson commented:
/test unit-tests

This looks related to #13636 , but not sure if that's just a bad merge by the automation.

skonto commented:
@evankanderson regarding b) it seems we are seeing failures because if containerPort is 0 (translates to random port) then it tries to add the NET_BIND_SERVICE. Downstream the [fix](https://github.com/openshift-knative/serving/pull/138/commits/2f0e76c71f9a163c9c5aaed727107045bea4602a) was to set specific port numbers outside that range.

evankanderson commented:
D'oh, I'd forgotten about the possibility of not setting a port number (which is pretty common) and then defaulting to zero.

I think we can simply adjust the test for port >0 and <1024.

Do you also want me to add something for opting out of `seccompProfile`?  It wouldn't be difficult, but it wasn't clear to me the affected range of components.

In any case, I think those fixes are for 1.10, as this feature is _disabled_ by default in in 1.9.

skonto commented:
> I think we can simply adjust the test for port >0 and <1024.
I think that should be the case.
> Do you also want me to add something for opting out of seccompProfile? It wouldn't be difficult, but it wasn't clear to me the affected range of components.
Yeah it would be nice because for OCP there are cases where you want to tag a service not to set seccompProfile because if set it does not allow you to run as the default image user (TestShouldRunAsUserContainerDefault). RuntimeDefault SeccompProfile on OCP brings restrictions that break stuff.

skonto commented:
> port >0 and <1024.

Verified it does the trick btw.
